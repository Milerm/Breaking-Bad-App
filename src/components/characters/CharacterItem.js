import React from 'react'
import { Card, Col, Spinner } from 'react-bootstrap'
import { Link } from 'react-router-dom'
import { useFetch } from '../../hooks/useFetch'
import PropTypes from 'prop-types'

export const CharacterItem = ({ character }) => {
  const { data, loading, hasError } = useFetch(`/quote?author=${character.name}`)

  return (
    <Col>
      <Card className="h-100p">
        <Card.Img variant="top" src={character.img} />
        <Card.Body>
          <Card.Title>{character.name}</Card.Title>
          <Card.Text as="div">
            {loading ? (
              <Spinner animation="border" variant="primary" />
            ) : hasError ? (
              <span>Error al obtener los datos</span>
            ) : data && data.length > 0 ? (
              <ul>
                {data.map((item) => (
                  <li key={item.quote_id}>
                    <Link to={`/quote/${item.quote_id}`}>{item.quote}</Link>
                  </li>
                ))}
              </ul>
            ) : (
              <p>No tenemos frases disponibles</p>
            )}
          </Card.Text>
        </Card.Body>
      </Card>
    </Col>
  )
}

CharacterItem.propTypes = {
  character: PropTypes.object.isRequired
}
