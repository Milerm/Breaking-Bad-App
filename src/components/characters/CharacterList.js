import React, { useEffect, useState } from 'react'
import { Button, Col, Container, Form, Pagination, Row } from 'react-bootstrap'
import { fetchWithoutToken } from '../../helpers/fetch'
import Util from '../../helpers/util'
import { useFetch } from '../../hooks/useFetch'
import { useForm } from '../../hooks/useForm'
import { CharacterItem } from './CharacterItem'

const initialForm = {
  category: 'category',
  text: ''
}

export const CharacterList = () => {
  const limit = 5
  const [{ category, text }, handleInputChange, reset] = useForm(initialForm)
  const [page, setPage] = useState(1)
  const [queryParams, setQueryParams] = useState('?')

  let { data, loading, hasError, refetch } = useFetch(
    `/characters${queryParams}limit=${limit}&offset=${(page - 1) * limit}`
  )

  let { data: allCharacters } = useFetch(`/characters${queryParams}`)
  const [characters, setCharacters] = useState([])
  const [listCharacters, setListCharacters] = useState([])
  const [paginationItems, setPaginationItems] = useState([])
  const [maxPages, setMaxPages] = useState(0)

  const changePage = async (nextPage) => {
    setPage(nextPage)
    await refresh(nextPage)
  }

  const handleSearch = async (e) => {
    e.preventDefault()
    setPage(1)
    await refresh()
  }

  const refresh = async (nextPage) => {
    const query = !Util.isNullOrUndefinedOrEmpty(text) ? `?${category}=${text}&` : '?'
    setQueryParams(query)
    let response = await fetchWithoutToken(
      `characters${query}limit=${limit}&offset=${(nextPage - 1) * limit}`
    )
    let body = await response.json()
    setCharacters(body)

    response = await fetchWithoutToken(`characters${query}`)
    body = await response.json()
    setListCharacters(body)
  }

  useEffect(() => {
    setCharacters(data)
    setListCharacters(allCharacters)
  }, [data, allCharacters])

  useEffect(() => {
    let pagItems = []
    const mPages = Math.ceil(listCharacters?.length / limit)
    for (let number = 1; number <= mPages; number++) {
      pagItems.push(
        <Pagination.Item key={number} onClick={() => changePage(number)} active={number === page}>
          {number}
        </Pagination.Item>
      )
    }
    setMaxPages(mPages)
    setPaginationItems(pagItems)
  }, [listCharacters])

  return (
    <div>
      <h1>Characters</h1>
      <Container fluid>
        <Row xs={1}>
          <Pagination className="pagination-center">
            <Pagination.First onClick={() => changePage(1)} disabled={page === 1} />
            <Pagination.Prev onClick={() => changePage(page - 1)} disabled={page === 1} />
            {paginationItems}
            <Pagination.Next onClick={() => changePage(page + 1)} disabled={page === maxPages} />
            <Pagination.Last onClick={() => changePage(maxPages)} disabled={page === maxPages} />
          </Pagination>
        </Row>
        <Form onSubmit={handleSearch}>
          <Form.Row>
            <Form.Group as={Col}>
              <Form.Control
                as="select"
                name="category"
                onChange={handleInputChange}
                value={category}
                custom
              >
                <option value="category">Category</option>
                <option value="name">Name</option>
              </Form.Control>
            </Form.Group>

            <Form.Group as={Col}>
              <Form.Control
                name="text"
                onChange={handleInputChange}
                value={text}
                placeholder="Ingrese texto a buscar..."
              />
            </Form.Group>

            <Form.Group as={Col}>
              <Button type="submit">
                <i className="fas fa-search" />
              </Button>
            </Form.Group>
          </Form.Row>
        </Form>
        <Row xs={2} sm={3} md={4} lg={5}>
          {loading ? (
            <span>Buscando personajes...</span>
          ) : hasError ? (
            <span>Error al obtener los datos</span>
          ) : characters && characters.length > 0 ? (
            characters.map((item) => <CharacterItem key={item.char_id} character={item} />)
          ) : (
            <p>No tenemos personajes disponibles</p>
          )}
        </Row>
      </Container>
    </div>
  )
}
