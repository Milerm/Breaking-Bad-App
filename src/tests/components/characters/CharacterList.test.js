import React from 'react'
import { shallow } from 'enzyme'
import { CharacterList } from '../../../components/characters/CharacterList'

describe('Pruebas en <CharacterList />', () => {
  const wrapper = shallow(<CharacterList />)

  test('debe mostrar el componente correctamente', () => {
    expect(wrapper).toMatchSnapshot()
  })

  // test('debe listar los primeros 5 personajes', () => {
  //   const items = wrapper.find('CharacterItem').length
  //   expect(items).toBe(5)
  // })
  
})
