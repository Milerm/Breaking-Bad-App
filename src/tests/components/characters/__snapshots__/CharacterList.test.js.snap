// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <CharacterList /> debe mostrar el componente correctamente 1`] = `
<div>
  <h1>
    Characters
  </h1>
  <Container
    fluid={true}
  >
    <Row
      noGutters={false}
      xs={1}
    >
      <ForwardRef
        className="pagination-center"
      >
        <First
          disabled={true}
          onClick={[Function]}
        />
        <Prev
          disabled={true}
          onClick={[Function]}
        />
        <Next
          disabled={false}
          onClick={[Function]}
        />
        <Last
          disabled={false}
          onClick={[Function]}
        />
      </ForwardRef>
    </Row>
    <Row
      lg={5}
      md={4}
      noGutters={false}
      sm={3}
      xs={2}
    >
      <p>
        No tenemos personajes disponibles
      </p>
    </Row>
  </Container>
</div>
`;
